### PR TITLE
Update Pregnancy Outcome Event Obs with multiple babies repeat group values list 

### PR DESCRIPTION
**Pull request by allan-on**
- Added Util to build a values map from a repeating group field
- Updated `save` pregnancy outcome Event
  - Added updating Event with Obs from the above-mentioned map. The Obs values are a JSON string of human-readable values from the repeating group in the order expected

----
See: 
- https://github.com/opensrp/opensrp-client-chw/issues/1974 
For details

### CODE REVIEW

File: opensrp-chw-anc/src/main/java/org/smartregister/chw/anc/interactor/BaseAncRegisterInteractor.java [hunk 1]
```diff
@@ -54,6 +54,51 @@ public BaseAncRegisterInteractor() {
...
+    public static void saveVaccineEvents(JSONArray fields, String baseID, String dob) {
+
+        JSONObject vaccinesAtBirthObject = org.smartregister.util.JsonFormUtils.getFieldJSONObject(fields, DBConstants.KEY.VACCINES_AT_BIRTH);
+        JSONArray vaccinesAtBirthArray = vaccinesAtBirthObject != null ? vaccinesAtBirthObject.optJSONArray(DBConstants.KEY.OPTIONS) : null;
+
+        if (vaccinesAtBirthArray != null) {
+            for (int i = 0; i < vaccinesAtBirthArray.length(); i++) {
+                JSONObject currentVaccine = vaccinesAtBirthArray.optJSONObject(i);
+                if (currentVaccine != null && currentVaccine.optBoolean(JsonFormUtils.VALUE) && !currentVaccine.optString(JsonFormUtils.KEY).equals("chk_none")) {
+                    String VaccineName = currentVaccine.optString(JsonFormUtils.KEY).equals("chk_bcg") ? "bcg" : "opv_0";
```
Comment (bennsimon) [inline opensrp-chw-anc/src/main/java/org/smartregister/chw/anc/interactor/BaseAncRegisterInteractor.java:66]: Could you move the string literals to constants class.

File: opensrp-chw-anc/src/main/java/org/smartregister/chw/anc/interactor/BaseAncRegisterInteractor.java [hunk 2]
```diff
@@ -54,6 +54,51 @@ public BaseAncRegisterInteractor() {
         this(new AppExecutors());
     }
 
+    public static void saveVaccineEvents(JSONArray fields, String baseID, String dob) {
```
Comment (bennsimon) [inline opensrp-chw-anc/src/main/java/org/smartregister/chw/anc/interactor/BaseAncRegisterInteractor.java:57]:
```suggestion
    public static void saveVaccineEvents(JSONArray fields, String baseId, String dob) {
```

File: opensrp-chw-anc/src/main/java/org/smartregister/chw/anc/interactor/BaseAncRegisterInteractor.java [hunk 3]
```diff
@@ -54,6 +54,51 @@ public BaseAncRegisterInteractor() {
...
+        try {
+            return formatter.parse(vaccineDate);
+        } catch (ParseException e) {
+            Timber.e(e);
+        }
+        return null;
+    }
+
+    private static void saveVaccines(JSONArray fields, String baseID) {
+        String[] vaccines = {"bcg_date", "opv0_date"};
```
Comment (bennsimon) [inline opensrp-chw-anc/src/main/java/org/smartregister/chw/anc/interactor/BaseAncRegisterInteractor.java:87]: Move string literals to constants class.

File: opensrp-chw-anc/src/main/java/org/smartregister/chw/anc/interactor/BaseAncRegisterInteractor.java [hunk 4]
```diff
@@ -54,6 +54,51 @@ public BaseAncRegisterInteractor() {
...
+        for (int i = 0; i < vaccines.length; i++) {
+            try {
+                String vaccineDate = getFieldJSONObject(fields, vaccines[i]).optString(VALUE);
+                if (StringUtils.isNotBlank(vaccineDate)) {
+                    Date dateVaccinated = vaccinationDate(vaccineDate);
+                    String VaccineName = vaccines[i] == "bcg_date" ? "bcg" : "opv_0";
```
Comment (bennsimon) [inline opensrp-chw-anc/src/main/java/org/smartregister/chw/anc/interactor/BaseAncRegisterInteractor.java:93]: Move string literals to constants class.

File: opensrp-chw-anc/src/main/java/org/smartregister/chw/anc/interactor/BaseAncRegisterInteractor.java [hunk 5]
```diff
@@ -54,6 +54,51 @@ public BaseAncRegisterInteractor() {
...
+                    VisitUtils.savePncChildVaccines(VaccineName, baseID, dateVaccinated);
+                }
+            } catch (NullPointerException e) {
+                Timber.d(e);
```
Comment (bennsimon) [inline opensrp-chw-anc/src/main/java/org/smartregister/chw/anc/interactor/BaseAncRegisterInteractor.java:97]: Should this be debug or error?

File: opensrp-chw-anc/src/main/java/org/smartregister/chw/anc/interactor/BaseAncRegisterInteractor.java (hunk 1 above)
Comment (bennsimon) [inline opensrp-chw-anc/src/main/java/org/smartregister/chw/anc/interactor/BaseAncRegisterInteractor.java:66]:
```suggestion
                    String vaccineName = currentVaccine.optString(JsonFormUtils.KEY).equals("chk_bcg") ? "bcg" : "opv_0";
```

File: opensrp-chw-anc/src/main/java/org/smartregister/chw/anc/interactor/BaseAncRegisterInteractor.java (hunk 4 above)
Comment (bennsimon) [inline opensrp-chw-anc/src/main/java/org/smartregister/chw/anc/interactor/BaseAncRegisterInteractor.java:93]:
```suggestion
                    String vaccineName = vaccines[i] == "bcg_date" ? "bcg" : "opv_0";
```

File: opensrp-chw-anc/src/main/java/org/smartregister/chw/anc/util/JsonFormUtils.java [hunk 6]
```diff
@@ -218,6 +223,30 @@ public static String getValue(JSONObject jsonObject, String key) {
...
+     * Returns an Array value from JSON form field
+     *
+     * @param jsonObject native forms jsonObject
+     * @param key        field object key
+     * @return Array value
+     */
+    public static JSONArray getValueArray(JSONObject jsonObject, String key) {
+        try {
+            JSONArray jsonArray = jsonObject.getJSONObject(JsonFormConstants.STEP1).getJSONArray(JsonFormConstants.FIELDS);
+            int x = 0;
```
Comment (bennsimon) [inline opensrp-chw-anc/src/main/java/org/smartregister/chw/anc/util/JsonFormUtils.java:236]: Could you rename `x` to something more descriptive?

File: opensrp-chw-anc/src/main/java/org/smartregister/chw/anc/util/JsonFormUtils.java [hunk 7]
```diff
@@ -218,6 +223,30 @@ public static String getValue(JSONObject jsonObject, String key) {
...
+            while (jsonArray.length() > x) {
+                JSONObject jo = jsonArray.getJSONObject(x);
```
Comment (bennsimon) [inline opensrp-chw-anc/src/main/java/org/smartregister/chw/anc/util/JsonFormUtils.java:238]: Could you rename `jo` to something more descriptive?

File: opensrp-chw-anc/src/main/java/org/smartregister/chw/anc/util/JsonFormUtils.java [hunk 8]
```diff
@@ -218,6 +223,30 @@ public static String getValue(JSONObject jsonObject, String key) {
...
     }
 
+    /**
+     * Returns an Array value from JSON form field
+     *
+     * @param jsonObject native forms jsonObject
+     * @param key        field object key
+     * @return Array value
+     */
+    public static JSONArray getValueArray(JSONObject jsonObject, String key) {
```
Comment (bennsimon) [inline opensrp-chw-anc/src/main/java/org/smartregister/chw/anc/util/JsonFormUtils.java:233]: Is this method meant to only work with single step form?

File: opensrp-chw-anc/src/main/java/org/smartregister/chw/anc/util/JsonFormUtils.java [hunk 9]
```diff
@@ -441,6 +470,86 @@ public static void updateFormField(JSONArray formFieldArrays, String formFieldKe
...
+                for (String key : repeatingGroupsMap.keySet()) {
+                    repeatArray.put(new JSONObject(Objects.requireNonNull(repeatingGroupsMap.get(key))));
+                }
+                List<Object> obsValues = new ArrayList<>();
+                obsValues.add(repeatArray.toString());
+                Obs formattedRepeatObs = new Obs();
+                formattedRepeatObs.setFieldCode("child_repeat_group_values_list");
+                formattedRepeatObs.setFieldDataType("repeatvalueslist");
+                formattedRepeatObs.setFieldType("concept");
+                formattedRepeatObs.setFormSubmissionField("repeatvalueslist");
```
Comment (bennsimon) [inline opensrp-chw-anc/src/main/java/org/smartregister/chw/anc/util/JsonFormUtils.java:543]: Move the string literals to constant class.

**Review by bennsimon — CHANGES_REQUESTED** [review]

File: opensrp-chw-anc/src/main/java/org/smartregister/chw/anc/interactor/BaseAncRegisterInteractor.java (hunk 1 above)
Comment (allan-on) [reply opensrp-chw-anc/src/main/java/org/smartregister/chw/anc/interactor/BaseAncRegisterInteractor.java:66]: Done ✅

File: opensrp-chw-anc/src/main/java/org/smartregister/chw/anc/interactor/BaseAncRegisterInteractor.java (hunk 2 above)
Comment (allan-on) [reply opensrp-chw-anc/src/main/java/org/smartregister/chw/anc/interactor/BaseAncRegisterInteractor.java:57]: Refactored ✅

File: opensrp-chw-anc/src/main/java/org/smartregister/chw/anc/interactor/BaseAncRegisterInteractor.java (hunk 3 above)
Comment (allan-on) [reply opensrp-chw-anc/src/main/java/org/smartregister/chw/anc/interactor/BaseAncRegisterInteractor.java:87]: Done ✅

File: opensrp-chw-anc/src/main/java/org/smartregister/chw/anc/interactor/BaseAncRegisterInteractor.java (hunk 4 above)
Comment (allan-on) [reply opensrp-chw-anc/src/main/java/org/smartregister/chw/anc/interactor/BaseAncRegisterInteractor.java:93]: Done ✅

File: opensrp-chw-anc/src/main/java/org/smartregister/chw/anc/interactor/BaseAncRegisterInteractor.java (hunk 5 above)
Comment (allan-on) [reply opensrp-chw-anc/src/main/java/org/smartregister/chw/anc/interactor/BaseAncRegisterInteractor.java:97]: Fixed ✅

File: opensrp-chw-anc/src/main/java/org/smartregister/chw/anc/interactor/BaseAncRegisterInteractor.java (hunk 1 above)
Comment (allan-on) [reply opensrp-chw-anc/src/main/java/org/smartregister/chw/anc/interactor/BaseAncRegisterInteractor.java:66]: Refactored ✅

File: opensrp-chw-anc/src/main/java/org/smartregister/chw/anc/interactor/BaseAncRegisterInteractor.java (hunk 4 above)
Comment (allan-on) [reply opensrp-chw-anc/src/main/java/org/smartregister/chw/anc/interactor/BaseAncRegisterInteractor.java:93]: Refactored ✅

File: opensrp-chw-anc/src/main/java/org/smartregister/chw/anc/util/JsonFormUtils.java (hunk 6 above)
Comment (allan-on) [reply opensrp-chw-anc/src/main/java/org/smartregister/chw/anc/util/JsonFormUtils.java:236]: Refactored ✅

File: opensrp-chw-anc/src/main/java/org/smartregister/chw/anc/util/JsonFormUtils.java (hunk 7 above)
Comment (allan-on) [reply opensrp-chw-anc/src/main/java/org/smartregister/chw/anc/util/JsonFormUtils.java:238]: Refactored ✅

File: opensrp-chw-anc/src/main/java/org/smartregister/chw/anc/util/JsonFormUtils.java (hunk 8 above)
Comment (allan-on) [reply opensrp-chw-anc/src/main/java/org/smartregister/chw/anc/util/JsonFormUtils.java:233]: Refactored to handle a single step that's specified by passing the step name as a param ✅

File: opensrp-chw-anc/src/main/java/org/smartregister/chw/anc/util/JsonFormUtils.java (hunk 9 above)
Comment (allan-on) [reply opensrp-chw-anc/src/main/java/org/smartregister/chw/anc/util/JsonFormUtils.java]: Done ✅

File: opensrp-chw-anc/src/main/java/org/smartregister/chw/anc/interactor/BaseAncRegisterInteractor.java [hunk 10]
```diff
@@ -54,6 +59,51 @@ public BaseAncRegisterInteractor() {
...
+            Timber.e(e);
+        }
+        return null;
+    }
+
+    private static void saveVaccines(JSONArray fields, String baseId) {
+        String[] vaccineKeys = {BCG_DATE, OPV_0_DATE};
+        for (int i = 0; i < vaccineKeys.length; i++) {
+            try {
+                String vaccineDate = getFieldJSONObject(fields, vaccineKeys[i]).optString(VALUE);
```
Comment (bennsimon) [inline opensrp-chw-anc/src/main/java/org/smartregister/chw/anc/interactor/BaseAncRegisterInteractor.java:95]: Check for null on this method  `getFieldJSONObject(fields, vaccineKeys[i])`.

File: opensrp-chw-anc/src/main/java/org/smartregister/chw/anc/util/JsonFormUtils.java [hunk 11]
```diff
@@ -218,6 +225,32 @@ public static String getValue(JSONObject jsonObject, String key) {
...
+     * @param key        field object key
+     * @return Array value
+     */
+    public static JSONArray getValueArray(JSONObject formJsonObject, String step, String key) {
+        try {
+            JSONArray fieldsArray = formJsonObject.getJSONObject(step).getJSONArray(JsonFormConstants.FIELDS);
+            int totalFields = fieldsArray.length() ;
+            int objectIndex = 0;
+            while (objectIndex < totalFields) {
+                JSONObject jsonObject = fieldsArray.getJSONObject(objectIndex);
```
Comment (bennsimon) [inline opensrp-chw-anc/src/main/java/org/smartregister/chw/anc/util/JsonFormUtils.java:242]: Check for null on `jsonObject`.

File: opensrp-chw-anc/src/main/java/org/smartregister/chw/anc/util/JsonFormUtils.java [hunk 12]
```diff
@@ -218,6 +225,32 @@ public static String getValue(JSONObject jsonObject, String key) {
...
+     * Returns an Array value from JSON form field
+     *
+     * @param formJsonObject native forms jsonObject
+     * @param step       JSON Form step with the field
+     * @param key        field object key
+     * @return Array value
+     */
+    public static JSONArray getValueArray(JSONObject formJsonObject, String step, String key) {
+        try {
+            JSONArray fieldsArray = formJsonObject.getJSONObject(step).getJSONArray(JsonFormConstants.FIELDS);
```
Comment (bennsimon) [inline opensrp-chw-anc/src/main/java/org/smartregister/chw/anc/util/JsonFormUtils.java:238]: Check for null.

**Review by bennsimon — CHANGES_REQUESTED** [review]

File: opensrp-chw-anc/src/main/java/org/smartregister/chw/anc/util/JsonFormUtils.java (hunk 12 above)
Comment (allan-on) [reply opensrp-chw-anc/src/main/java/org/smartregister/chw/anc/util/JsonFormUtils.java:238]: This is handled in the try-catch that handles parsing the JSONObject and JSONArray

**Review by bennsimon — APPROVED** [review]

**Pull request merged**